drivers/mbox/mbox_nxp_imx_mu.c

```diff
@@ -118,66 +120,62 @@ static const struct mbox_driver_api nxp_imx_mu_driver_api = {
-	{									\
-		const struct device *dev = DEVICE_DT_INST_GET(idx);		\
-		const struct nxp_imx_mu_data *data = dev->data;			\
-		const struct nxp_imx_mu_config *config = dev->config;		\
-		struct mbox_msg msg;						\
-		struct mbox_msg *callback_msg_ptr = NULL;			\
-		uint32_t flag = MU_GetStatusFlags(config->base);		\
-										\
-		for (int i_channel = 0; i_channel < MU_MAX_CHANNELS; i_channel++) { \
-			if ((flag & (kMU_Rx0FullFlag >> i_channel)) == \
-				(kMU_Rx0FullFlag >> i_channel)) { \
-				mu_##idx##_received_data =				\
-					MU_ReceiveMsgNonBlocking(config->base, 0);	\
-				msg.data = (const void *)&mu_##idx##_received_data;	\
-				msg.size = MU_MBOX_SIZE;				\
-				callback_msg_ptr = &msg;				\
-			} else if ((flag & (kMU_GenInt0Flag >> i_channel)) == \
-						(kMU_GenInt0Flag >> i_channel)) {	\
-				MU_ClearStatusFlags(config->base, \
-					(kMU_GenInt0Flag >> i_channel)); \
-				callback_msg_ptr = NULL;				\
-			}								\
-											\
-			if (data->cb[i_channel]) {					\
-				data->cb[i_channel](dev, i_channel,				\
-						data->user_data[i_channel],		\
-						callback_msg_ptr);			\
-			} \
-		} \
+static void handle_irq(const struct device *dev);
+
+#define MU_INSTANCE_DEFINE(idx)                                                                    \
+	static struct nxp_imx_mu_data nxp_imx_mu_##idx##_data;                                     \
+	const static struct nxp_imx_mu_config nxp_imx_mu_##idx##_config = {                        \
+		.base = (MU_Type *)DT_INST_REG_ADDR(idx),                                          \
+	};                                                                                         \
+	void MU_##idx##_IRQHandler(void);                                                          \
+	static int nxp_imx_mu_##idx##_init(const struct device *dev)                               \
+	{                                                                                          \
+		ARG_UNUSED(dev);                                                                   \
+		MU_Init(nxp_imx_mu_##idx##_config.base);                                           \
+		IRQ_CONNECT(DT_INST_IRQN(idx), DT_INST_IRQ(idx, priority), MU_##idx##_IRQHandler,  \
+			    NULL, 0);                                                              \
+		irq_enable(DT_INST_IRQN(idx));                                                     \
+		return 0;                                                                          \
+	}                                                                                          \
+	DEVICE_DT_INST_DEFINE(idx, nxp_imx_mu_##idx##_init, NULL, &nxp_imx_mu_##idx##_data,        \
+			      &nxp_imx_mu_##idx##_config, POST_KERNEL, CONFIG_MBOX_INIT_PRIORITY,  \
+			      &nxp_imx_mu_driver_api)
+
+#define MU_IRQ_HANDLER(idx)                                                                        \
+	void MU_##idx##_IRQHandler(void)                                                           \
```

Comment:
This is a minor nit, but we don't strictly need this intermediate function definition. `IRQ_CONNECT` lets you set the argument passed to the ISR. See this driver for an example: https://github.com/zephyrproject-rtos/zephyr/blob/db429004fdea646289bdb778e40a171cacfd16cf/drivers/spi/spi_mcux_dspi.c#L205

https://github.com/zephyrproject-rtos/zephyr/blob/db429004fdea646289bdb778e40a171cacfd16cf/drivers/spi/spi_mcux_dspi.c#L927

Reply:
Thanks, noted for next time!